src/main/java/uk/gov/companieshouse/web/accounts/service/smallfull/impl/BalanceSheetServiceImpl.java

```diff
@@ -72,66 +83,152 @@ public BalanceSheet getBalanceSheet(String transactionId, String companyAccounts
+        return validationErrors;
+
+    }
 
+    private SmallFullApi getSmallFullData(ApiClient apiClient, String smallFullUri)
+            throws ServiceException {
         try {
-            apiClient.smallFull().currentPeriod().create(uri, currentPeriod).execute();
+            return apiClient.smallFull().get(smallFullUri).execute();
         } catch (ApiErrorResponseException e) {
+            throw new ServiceException("Error retrieving small full data", e);
+        } catch (URIValidationException e) {
+            throw new ServiceException("Invalid URI for small full resource", e);
+        }
+    }
+
+    private void createPreviousPeriod(ApiClient apiClient, SmallFullApi smallFullApi, String previousPeriodUri, PreviousPeriodApi previousPeriodApi, List<ValidationError> validationErrors)
+            throws ServiceException {
+        boolean isCreated = isPreviousPeriodCreated(smallFullApi.getLinks());
+        try {
+            if (!isCreated) {
+                apiClient.smallFull().previousPeriod().create(previousPeriodUri, previousPeriodApi).execute();
+            } else {
+                apiClient.smallFull().previousPeriod().update(previousPeriodUri, previousPeriodApi).execute();
+            }
 
+        } catch (ApiErrorResponseException e) {
             if (e.getStatusCode() == HttpStatus.BAD_REQUEST.value()) {
-                List<ValidationError> validationErrors = validationContext.getValidationErrors(e);
-                if (validationErrors == null) {
+                validationErrors.addAll(validationContext.getValidationErrors(e));
+                if (validationErrors.isEmpty()) {
                     throw new ServiceException("Bad request posting balance sheet", e);
                 }
+            }
+        } catch (URIValidationException e) {
+            throw new ServiceException("Invalid URI for previous period resource", e);
+        }
+    }
 
-                return validationErrors;
+    private void createCurrentPeriod(ApiClient apiClient, SmallFullApi smallFullApi, String currentPeriodUri, CurrentPeriodApi currentPeriod, List<ValidationError> validationErrors)
+            throws ServiceException {
+        boolean isCreated = isCurrentPeriodCreated(smallFullApi.getLinks());
+        try {
+            if (!isCreated) {
+                apiClient.smallFull().currentPeriod().create(currentPeriodUri, currentPeriod).execute();
+            } else {
+                apiClient.smallFull().currentPeriod().update(currentPeriodUri, currentPeriod).execute();
             }
 
-            throw new ServiceException("Error posting balance sheet", e);
+        } catch (ApiErrorResponseException e) {
+            if (e.getStatusCode() == HttpStatus.BAD_REQUEST.value()) {
+                validationErrors.addAll(validationContext.getValidationErrors(e));
+                if (validationErrors.isEmpty()) {
+                    throw new ServiceException("Bad request posting balance sheet", e);
```

Comment:
As above, the exception message here should be made more specific to differentiate it from a similar exception thrown in `createPreviousPeriod()`, for example: `Bad request when creating current period resource`.